app/src/main/java/de/westnordost/streetcomplete/data/osm/mapdata/MapDataCache.kt

```diff
@@ -64,12 +67,15 @@ class MapDataCache(
             spatialCache.update(deleted = deletedNodeIds)
             spatialCache.replaceAllInBBox(updatedNodes, bbox)
         }
+        // remove all cached nodes that are now in spatialCache
+        nodeCache.keys.removeAll { spatialCache.get(it) != null }
```

Comment:
I had checked this, see "update is typically around 10% slower". The 10% are caused by this and `update is typically around 10% slower`.
I can try using getKeys().toSet() instead and see whether there is a measureable difference.

Reply:
I wouldn't have expected only 10%. Apparently entering locking is faster than I thought.

Reply:
I did some test comparing it to a version of `spatialCache.get` without `synchronized`. Maybe there is a small difference (and there should be), but I couldn't find it just measuring the time.
In both cases it's about 5 ms for 4k nodes in `nodeCache` and 23k nodes in `spatialCache`

Reply:
Interesting. Well then, nevermind my comment then. Hm. Well, Copying 23k Longs first into a list and then into a set may also take some time...

Reply:
I didn't find much information I would definitely trust, but various claims:
* lock coarsening: thanks to optimizations `synchronized` section is entered only once
* `synchronized` is fast: some benchmarks claim between 4 and 15 ns, which would be less than 60 µs for 4k calls (on different hardware of course...)
* `synchronized` is not locking if it is used in a single thread (not sure if this would apply to `MapDataCache` even though everything in there is `synchronized`)

If at least one of those is correct, it would explain why there is no measurable difference.

Reply:
In case you have the sources that talk about these topics still at hand, I'd love to read them . Especially lock coarsening sounds interesting.

Reply:
I think it was mostly StackOverflow talk and benchmarks when I was searching about performance of synchronized.

> Especially lock coarsening sounds interesting.

Here I really can't help... I found it mentioned several times, but no information on what it actually does.